scalyr_agent/util.py

```diff
@@ -1257,6 +1257,25 @@ def charge_if_available(self, num_bytes, current_time=None):
 
         @return: True if there are enough room in the rate limit to allow the operation.
         """
+        if self._get_time_to_sleep(num_bytes, current_time) > 0.0:
+            return False
+        else:
+            self.__bucket_contents -= num_bytes
+            return True
+
+    def sleep_and_charge(self, num_bytes, current_time=None):
+        """Sleeps until there are enough bytes available for an operation costing num_bytes, then charges that amount.
+
+        @param num_bytes: The number of bytes to consume from the rate limit.
+        @param current_time: If not none, the value to use as the current time, expressed in seconds past epoch. This
+            is used in testing.
+        """
+        time_to_sleep = self._get_time_to_sleep(num_bytes, current_time)
+        if time_to_sleep > 0.0:
+            time.sleep(time_to_sleep)
```

Comment:
Technically, if someone sets a zero fill rate, we could try to sleep infinity.  I would instead check at the start of this method if fill rate is zero, and if so, throw an exception saying you cannot use this method if you have a zero fill rate.